keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,107 @@
+
+## Motivation
+
+The dependencies between KUDO, operators packages and the application that is managed by the operator enforce the use of multiple versions to fully describe these dependencies. First, there is the dependency on a specific API version of KUDO. Also, a specific version of an application is provided by the package. And last, the operator itself may have multiple revisions. All of these versions need to be captured as part of the operator in a concise way that makes it easy for users to install specific operators for any of these versions.
+
+### Goals
+
+* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
+* Allow users to install a specific version of an application and/or a specific revision of an operator
+* Reject packages that don't have a matching KUDO API version
+
+### Non-Goals
+
+* Handle operator dependencies on specific Kubernetes versions
+
+## Proposal
+
+A `operator.yaml` file already provides versions for all components describing an operator:
+  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
+  * The bundled application version is provided by the `appVersion` field
+  * The version of the operator is provided by the `version` field
+  * The version of Kubernetes is provided by the `kubernetesVersion` field
+
+From these versions, currently only the `version` field can be used when installing packages with `kubectl kudo install`. Furthermore, a package repository index only contains `appVersion` and `version` fields in its entries. This limits the possible installation scenarios.
+At least the following installation scenarios should be supported:
+  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
+  2. A user should be able to install a specific `appVersion` of an operator
+
+### Support multiple API versions in the package repository
+
+Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
+
+For this, the following changes have to be made in KUDO:
+
+* Add `apiVersion` to the package index by updating the `Metadata` struct in `pkg/kudoctl/util/repo` and filling these values in `pkg/kudoctl/util/repo.ToPackageVersion`. A `index.yaml` entry should then look like this:
+
+```
+  kafka:
+  - appVersion: 2.3.0
```

Comment:
we should clarify what will happen in case of a base tech + multiple base techs packed into one operator. 
in the case of Kafka, the mirror-maker and kafka manager, independent from the Apache Kafka project might land in the same operator. 
We will still have a versions matrix explaining that in operator version `0.2.0` we have kafka `2.3.0` kafka manager `2.0.0.2` and mirror maker `2.0.0` 
so giving preference to Kafka `2.3.0` as appVersion won't be close to the truth

Reply:
Good point! I looked into other operators in https://github.com/operator-framework/awesome-operators and saw similar "bundles" like what you're describing above.
I'd go so far as to make `appVersion` optional and using it as a helper for operator devs to indicate an application version _if they want to_. As it is free-form, it could be as complicated as "kafka-2.3.0_manager-2.0.0.2_mirror-2.0.0" if all these versions need to be captured.

Reply:
for the reference the latest Kafka operator versions:

strimzi 0.14.0
banzai kafka operator 0.8.0
yolean kafka operator 6.0.3

They all are offering Kafka `2.3.x`

Reply:
Another option would be to support both `appVersion` and `appVersions`. The latter being a list of app names and their versions, e.g.
```yaml
kafka:
  appVersions:
    - kafka: 2.3.0
    - kafka-manager: 2.0.0.2
    - mirror-make: 2.0.0
```
But this will make it complicated to allow users to select a specific app version when running `kubectl kudo install`.

Reply:
I think something like this:
`kubectl kudo install kafka` --> latest
`kubectl kudo install kafka --version=0.2.0` --> to install operator version 0.2.0
and use the kubernetes way of the selector for app versions
`kubectl kudo install kafka -l kafka=2.3.0 --version=0.2.0`
`kubectl kudo install kafka -l kafka=2.3.0,kafka-manager=2.0.0.2`
or selector
`kubectl kudo install kafka --selector kafka=2.3.0,kafka-manager=2.0.0.2,kudo=0.8.0` --> a version that has kafka 2.3.0, kakfa-manager 2.0.0.2 and can run on kudo 0.8.0

but this means putting app versions in the back seat as `labels` and provide users a way to install them but giving priority to the operator version by default.

Reply:
For "bundles", I see two alternatives that might make sense:
1. In the same way that some other non-KUDO operators hide the main underlying application version behind their own versions (as Zain illustrated), KUDO operators could also hide these non-main underlying application versions behind the KUDO version. For example, Kafka `2.3.0-1.0.0` will come with its additional underlying applications with given versions that are opaque from the perspective of the KUDO operator version itself.

2. Encourage separate applications to be separate KUDO operators and rely on KUDO to handle dependencies. For Elasticsearch for example, Elasticsearch would be an operator and Kibana would be a separate operator.

Reply:
These alternatives are not incompatible with each other, too.

Reply:
> we should clarify what will happen in case of a base tech + multiple base techs packed into one operator.

Ideally, this would be handled as dependencies and we wouldn't list these in the main version.

Reply:
I would be for a breaking change here and adding `versions` as an object of versions.   or `metadata1.   not critical.... but may have some readability value.

Reply:
> Ideally, this would be handled as dependencies

we cannot force that on the users who are planning to build app stacks on top of KUDO. In some certain cases for data-services, we might have one dominant base tech. But in the end it will be mostly a stack of apps bound together to serve a common domain. 

And we shouldn't bind current features with future features. Zookeeper is a dependency of Kafka but we still cannot use it as like that and specify which version of zookeeper it has to install.

Reply:
Another suggestion for bundles would be to use metadata to describe the versions of the underlying applications. With Kafka as example, users are _mostly_ interested in running a specific version of Kafka, not so much of the bundled applications. Of course, this changes if there are bugs in the bundled apps and a user wants to be sure that these are fixed. By having `appVersion` describe the version of Kafka or the whole bundle, the main use case is covered, i.e. user wanting to install a specific version of Kafka. By having the versions of the bundled applications provided as metadata, users dependent on these versions can manually verify that they aren't affected by certain issues with these applications.

Reply:
I have a feeling here we can hash out in more detail (and I've missed some convos thanks to Kubecon so I'm definitely coming back up to speed), so correct me if I'm off base or this has already been discussed.

I'm concerned with mixing in too many base app versions. I hypothesize that people are installing Kafka 2.3.0, and the ecosystem of tooling is just an implementation detail for the operator itself. In this world, a change to mirror maker, kafka manager, etc., would be an _operator version_ revision, not an increase on some appVersion matrix. It's a simplified (in the simplex / less resolution) model that doesn't capture all of the semantics, but I think this is OK versus trying to capture everything possible.

Other than that, I echo what @nfnt's said -- which is mostly the same. :)

Reply:
Also, if there's truly two top level base tech versions here, then you should have two operators with dependencies, not a single operator with multiple base tech versions.

Combined/stack operators (say, a data processing stack) still present a problem here that we need to work around, of course, but I'd even consider solving that with another CRD that has this list-of-versions semantic since it is inherently different.

Reply:
I think we have two main use cases:
- Single Application Operator -> appVersion is simple
- Bundle Operators
   -- Has a "main" application (i.e. Kafka, plus some tooling) -> appVersion should be based on the main
  -- No "main" application (i.e. some encapsulated domain specific bundle) -> In this case the appVersion should simply not set and the operator just has the operator version and the developer has to handle the underlying versioning herself.